### PR TITLE
Report node out of disk condition and use that information to schedule pods.

### PR DESCRIPTION
**Pull request by madhusudancs**
Implements changes proposed in issue #4135 

David had already reviewed this change once before. I am doing push -f. So here is the reference to the reviewed commit - https://github.com/madhusudancs/kubernetes/commit/71a528a4f1bceea79a86c8eeca9e2d9ba77fadd9

### CODE REVIEW

**Comment by madhusudancs** [comment]
I forgot to mention that I am still working on an e2e test for this PR.

File: pkg/client/cache/listers.go [hunk 1]
```diff
@@ -95,6 +95,10 @@ func (s *StoreToPodLister) Exists(pod *api.Pod) (bool, error) {
 	return exists, nil
 }
 
+// NodeConditionPredicate is a function that indicates if a given list of node conditions match the
```
Comment (davidopp) [inline pkg/client/cache/listers.go]: This comment is confusing because the argument to the function is not a list of node conditions. I would slightly re-word to say "NodeConditionPredicate is a function that indicates whether the given node's conditions meet some set of criteria defined by the function."

File: pkg/client/cache/listers.go [hunk 2]
```diff
@@ -109,48 +113,26 @@ func (s *StoreToNodeLister) List() (machines api.NodeList, err error) {
...
-			if nodeCondition.Status == s.conditionStatus {
-				nodes.Items = append(nodes.Items, node)
-			} else {
-				glog.V(4).Infof("Ignoring node %v with condition status %v", node.Name, nodeCondition.Status)
-			}
+		if s.predicate(node) {
+			nodes.Items = append(nodes.Items, node)
 		} else {
-			glog.V(2).Infof("Node %s doesn't have conditions of type %v", node.Name, s.conditionType)
+			glog.V(2).Infof("Node %s doesn't match any condition", node.Name)
```
Comment (davidopp) [inline pkg/client/cache/listers.go]: s/any condition/all conditions/

File: pkg/controller/node/nodecontroller.go [hunk 3]
```diff
@@ -532,6 +532,32 @@ func (nc *NodeController) tryUpdateNodeStatus(node *api.Node) (time.Duration, ap
 				readyCondition.LastTransitionTime = nc.now()
 			}
 		}
+
+		// Like NodeReady condition, NodeOutOfDisk was last set longer ago than gracePeriod, so update
+		// it to Unknown (regardless of its current value) in the master, without contacting kubelet.
```
Comment (davidopp) [inline pkg/controller/node/nodecontroller.go]: The "without contacting kubelet" thing is a hold-over from a previous implementation that we no longer use. Please delete that phrase from this comment and the one above.

File: pkg/controller/node/nodecontroller.go [hunk 4]
```diff
@@ -532,6 +532,32 @@ func (nc *NodeController) tryUpdateNodeStatus(node *api.Node) (time.Duration, ap
...
+		oodCondition := nc.getCondition(&node.Status, api.NodeOutOfDisk)
```
Comment (davidopp) [inline pkg/controller/node/nodecontroller.go]: Please add a TODO to refactor this with readyCondition to remove duplicated code

File: pkg/controller/node/nodecontroller.go [hunk 5]
```diff
@@ -532,6 +532,32 @@ func (nc *NodeController) tryUpdateNodeStatus(node *api.Node) (time.Duration, ap
...
+		if oodCondition == nil {
+			glog.V(2).Infof("Out of disk condition of node %v is never updated by kubelet", node.Name)
+			oodCondition := &api.NodeCondition{
```
Comment (davidopp) [inline pkg/controller/node/nodecontroller.go]: nit: Is there any reason to do it this way instead of the way it was done for readyCondition, namely
node.Status.Conditions = append(node.Status.Conditions, api.NodeCondition{ ...}) ?
If not, I'd suggesting mirroring the way it was done for readyCondition, just to make it clearer to the reader that what you did is exactly the same logic.

File: pkg/controller/node/nodecontroller.go [hunk 6]
```diff
@@ -532,6 +532,32 @@ func (nc *NodeController) tryUpdateNodeStatus(node *api.Node) (time.Duration, ap
...
+			}
+			node.Status.Conditions = append(node.Status.Conditions, *oodCondition)
+		} else {
+			glog.V(2).Infof("node %v hasn't been updated for %+v. Last out of disk condition is: %+v",
+				node.Name, nc.now().Time.Sub(savedNodeStatus.probeTimestamp.Time), oodCondition)
+			if oodCondition.Status != api.ConditionUnknown {
+				oodCondition.Status = api.ConditionUnknown
+				oodCondition.Reason = "NodeStatusUnknown"
+				oodCondition.Message = fmt.Sprintf("Kubelet stopped posting node status.")
+				oodCondition.LastTransitionTime = nc.now()
```
Comment (davidopp) [inline pkg/controller/node/nodecontroller.go]: Is there a reason you didn't set oodCondition.LastHeartbeatTime? I think you should set it, even though some of this is redundant with readyCondition.

File: pkg/kubelet/kubelet.go [hunk 7]
```diff
@@ -2457,6 +2457,61 @@ func (kl *Kubelet) setNodeStatus(node *api.Node) error {
...
+	// is created its status is set to api.ConditionUnknown which matches either
+	// nodeOODCondition.Status != api.ConditionTrue or
+	// nodeOODCondition.Status != api.ConditionFalse in the conditions below depending on whether
+	// the kubelet is out of disk or not.
+	if kl.isOutOfDisk() {
+		if nodeOODCondition.Status != api.ConditionTrue {
+			nodeOODCondition.Status = api.ConditionTrue
+			nodeOODCondition.Reason = "KubeletOutOfDisk"
+			nodeOODCondition.Message = "out of disk space"
+			nodeOODCondition.LastTransitionTime = currentTime
```
Comment (davidopp) [inline pkg/kubelet/kubelet.go]: Is there some reason why you don't also set LastHeartbeatTime here? (to currentTime)

File: pkg/kubelet/kubelet.go [hunk 8]
```diff
@@ -2457,6 +2457,61 @@ func (kl *Kubelet) setNodeStatus(node *api.Node) error {
...
+			kl.recordNodeStatusEvent("NodeOutOfDisk")
+		}
+	} else {
+		if nodeOODCondition.Status != api.ConditionFalse {
+			nodeOODCondition.Status = api.ConditionFalse
+			nodeOODCondition.Reason = "KubeletHasSufficientDisk"
+			nodeOODCondition.Message = "kubelet has sufficient disk space available"
+			nodeOODCondition.LastTransitionTime = currentTime
```
Comment (davidopp) [inline pkg/kubelet/kubelet.go]: Is there some reason why you don't also set LastHeartbeatTime here? (to currentTime)

File: pkg/kubelet/kubelet_test.go [hunk 9]
```diff
@@ -2666,6 +2728,21 @@ func TestUpdateNodeStatusWithoutContainerRuntime(t *testing.T) {
 	}
 	mockCadvisor.On("VersionInfo").Return(versionInfo, nil)
 
+	// Create a new DiskSpaceManager with a new policy. This new manager along with the mock FsInfo
+	// values added to Cadvisor should make the kubelet report that it is out of disk space.
```
Comment (davidopp) [inline pkg/kubelet/kubelet_test.go]: I think this should say "it has sufficient disk space" ?

**Comment by davidopp** [comment]
Very nice work!

Jut a few small comments. (A few of my comments you may have seen in email were actually wrong; I've deleted those from Github.)

@lavalamp in case he wants to make a quick pass over pkg/client/cache/listers.go

@dchen1107 @yujuhong in case either of them wants to make a quick pass over kubelet.go

(I reviewed the whole thing, so it's not strictly necessary, but thought I'd give a heads-up.)

**Comment by yujuhong** [comment]
> @dchen1107 @yujuhong in case either of them wants to make a quick pass over kubelet.go
> 
> (I reviewed the whole thing, so it's not strictly necessary, but thought I'd give a heads-up.)

Thanks for the heads up. lgtm.

File: pkg/kubelet/kubelet_test.go (hunk 9 above)
Comment (madhusudancs) [inline pkg/kubelet/kubelet_test.go]: Thanks for catching that. It's a copy-paste fail. It is probably also an indication that this should be extracted out to a function. I will leave it as is for now. I will extract it out to a function if I make any more modifications.

File: pkg/controller/node/nodecontroller.go (hunk 5 above)
Comment (madhusudancs) [inline pkg/controller/node/nodecontroller.go]: There is no good reason now to do this in a different way. I tried to do something to make this logic clear, but abandoned that approach. This difference is a left over from what I was trying to do.

Reverting to make this look like readyCondition.

File: pkg/controller/node/nodecontroller.go (hunk 4 above)
Comment (madhusudancs) [inline pkg/controller/node/nodecontroller.go]: Done.

File: pkg/controller/node/nodecontroller.go (hunk 3 above)
Comment (madhusudancs) [inline pkg/controller/node/nodecontroller.go]: Done.

File: pkg/controller/node/nodecontroller.go (hunk 5 above)
Comment (yujuhong) [inline pkg/controller/node/nodecontroller.go]: IMO, we should refactor and move all node status code out of kubelet.go. This should include factor out common code for setting conditions. That deserves its own PR though.

File: pkg/client/cache/listers.go (hunk 2 above)
Comment (madhusudancs) [inline pkg/client/cache/listers.go]: "doesn't match all conditions" will sound a little odd. How about "Node %s matches none of the conditions"?

File: pkg/kubelet/kubelet.go [hunk 10]
```diff
@@ -2457,6 +2457,61 @@ func (kl *Kubelet) setNodeStatus(node *api.Node) error {
 			kl.recordNodeStatusEvent("NodeNotReady")
 		}
 	}
+
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: IMO, we should refactor and move all node status code out of kubelet.go. This should include factor out common code for setting conditions. The pattern here is hard to follow. That deserves its own PR though.

Comment (madhusudancs) [inline pkg/kubelet/kubelet.go]: I definitely agree. The pattern and logic here is hard to follow. I spent my entire first week just trying to understand the pattern and logic here. Also, the fact that this file and its accompanying test are gigantic aggravates things.

Added a TODO.

File: pkg/client/cache/listers.go (hunk 1 above)
Comment (madhusudancs) [inline pkg/client/cache/listers.go]: Done.

Thanks for that wording. I was having trouble to make this comment more readable.

**Comment by madhusudancs** [comment]
This PR is now superseded by PRs #16178 and #16179. Abandoning this PR. 

@davidopp sent a note explaining why this is necessary. Pasting his note verbatim:

> Hi Madhu. I talked to Dawn and Brian about whether we should cherrypick your OutOfDisk PR into 1.1 The consensus was that it's too risky to cherrypick to whole thing, but we were wondering if you could split it up into two PRs, one with just the Kubelet change (and everything it depends on and the associated tests) and one with everything else? Then we could cherrypick the first PR into the release. It will still be helpful because the OutOfDisk condition can be used for debugging.
